server/controllers/applications.coffee

```diff
@@ -1,20 +1,36 @@
 appManager = require '../lib/app_manager'
 {getProxy} = require '../lib/proxy'
-
+fs = require 'fs'
 
 module.exports.app = (req, res, next) ->
     appName = req.params.name
     req.url = req.url.substring "/apps/#{appName}".length
     shouldStart = -1 is req.url.indexOf 'socket.io'
-    appManager.ensureStarted appName, shouldStart, (err, port) ->
+    appManager.ensureStarted appName, shouldStart, (err, result) ->
         if err?
             error = new Error err.msg
             error.status = err.code
             error.template =
                 name: if err.code is 404 then 'not_found' else 'error_app'
             next error
+        else if result.port?
+            getProxy().web req, res, target: "http://localhost:#{result.port}"
         else
-            getProxy().web req, res, target: "http://localhost:#{port}"
+            # if public repository exists, then redirect to public app
+            if fs.existsSync result.path + '/public/index.html'
+                req.url = "/public/#{appName}"
+                res.redirect "#{req.url}/"
+            else
+                # showing private static app
+                appManager.startStaticApp appName, req.url, result.path, (
```

Comment:
It would be better to use one of the many static assets handler middleware, or even res.sendFile. 
It avoid loading the whole index.html into memory, better manage request aborting and for the middleware case, can handle caching too.